### PR TITLE
[WIP] Wallet: Cache CWalletDB pointer in CWallet to improve performance

### PR DESCRIPTION
**Pull request by pstratem**

### CODE REVIEW

**Comment by jonasschnelli** [comment]
Concept ACK.
Travis repots issue with rpc tests. Keeping the DB connection permanently open – I agree – is the way to go. Although it might introduce new bugs.

**Comment by pstratem** [comment]
@jonasschnelli this is very very much a WIP :)

**Comment by jonasschnelli** [comment]
Right. I almost oversaw the WIP tag.
Nice work btw!

**Comment by laanwj** [comment]
Have you measured the performance improvements?

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -274,7 +274,7 @@ bool CWallet::ChangeWalletPassphrase(const SecureString& strOldWalletPassphrase,
                     return false;
                 if (!crypter.Encrypt(vMasterKey, pMasterKey.second.vchCryptedKey))
                     return false;
-                CWalletDB(strWalletFile).WriteMasterKey(pMasterKey.first, pMasterKey.second);
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: There is a reason we create CWalletDB objects every time per operation. From what I remember this assures that the transactions are closed/flushed properly.

**Comment by pstratem** [comment]
@laanwj I have not specifically but "much much much faster" is how i would describe it

**Comment by laanwj** [comment]
OK. That's good to hear. Just need to verify that it doesn't come at the expense of robustness, then.

**Comment by pstratem** [comment]
@laanwj CDB::~CDB simply calls CDB::Close which basically just calls Flush

So it's simply a matter of ensuring that we call Flush properly

**Comment by pstratem** [comment]
The easier thing to do is to simple change the default for fFlushOnClose to false, but that doesn't allow for optimizing batch operations such as keypoolrefill (which was actually my original motivation for this change).

**Comment by pstratem** [comment]
Will do this more incrementally.

**Pull request closed without merging**